13_Multiple-testing.Rmd

```diff
@@ -631,6 +640,332 @@ e.g. the `samr` package in `R`.
+
+Compare the FDR results for two methods.
+
+1.  First method: using `p.adjust()` function from {stats} package
+```{r}
+pvals = results$p.value
+pvals<-sort(pvals)
+
+# to find a q-value with the false discovery rate method
+fdr <- p.adjust(pvals, method="fdr")
+
+sum(fdr<0.05)
+```
+```{r}
+mean(fdr<0.05)
+```
+
+2.  Second method: using the `qvalue()` function from {qvalue} package
+```{r}
+res <- qvalue::qvalue(pvals)
+qvals <- res$qvalues
+#plot(pvals,qvals)
+sum(qvals<0.05)
+```
+```{r}
+mean(qvals<0.05)
+```
+
+The proportion of true null hypotheses:
+```{r}
+res$pi0
+```
+
+
+```{r eval=FALSE, include=FALSE}
+plot(qvals/fdr)
+abline(h=res$pi0,col=2)
+```
+
+
+```{r}
+hist(pvals,breaks=seq(0,1,len=21))
+expectedfreq <- length(pvals)/20 #per bin
+abline(h=expectedfreq*qvalue(pvals)$pi0,col=2,lty=2)
+```
+
+## Replications
+
+Little recap: **There is a trade-off between type I & type II error**
+
+- **Type I error** reject $H_0$ when $H_0$ is TRUE
+- **Type I error Rate** is the prob of type I error
+- **Type II error** no reject $H_0$ when $H_0$ is FALSE
+- **POWER** of hypothesis is the prob of not making type II error
+
+
+
+![](images/13_03.png){width=80%}
+
+![credit: Federica Gazzelloni](images/13_mht_tbl.png){width=80%}
+
+
+
+- R: sum of the number of pvalues which are below the threshold and will be rejected (n. rejections)
+- m: total hypothesis testing
+- m0: negatives
+- m1: m-m0 positives
+
+Now what we do is replicating the sample in a lab environment to obtain fake data, in order to do that we generate a matrix from a replication of normal distribution of the same size of our data.
+
+```{r fake}
+n <- 24
+m <- 8793
+delta <- 2 # this is the sample split in two
+positives <- 500 # m1
+# negatives
+m0<- m-positives # 8293
+
+mat <- matrix(rnorm(n*m),m,n)
+mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)] + delta
+```
+
+
+Then we do 1000 replications to see in what is the false discovery rate for this lab.
+```{r fake2, eval=FALSE, include=T}
+B<-1000
+set.seed(1173)
+results_global <- replicate(B,{
+  mat <- matrix(rnorm(n*m),m,n)
+  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)] + delta
+  pvals = genefilter::rowttests(mat,null_hypothesis)$p.val
+  ##Bonferroni
+  FP1 <- sum(pvals[-(1:positives)]<=0.05/m)  
+  FN1 <- sum(pvals[1:positives]>0.05/m)
+  # p.adjust
+  qvals1 <- p.adjust(pvals,method = "fdr")
+  FP2<-sum(qvals1[-(1:positives)]<=0.05)
+  FN2 <- sum(qvals1[1:positives]>0.05)
+  # qvalue
+  qvals2 <- qvalue::qvalue(pvals)$qvalues
+  FP3<-sum(qvals2[-(1:positives)]<=0.05)
+  FN3 <- sum(qvals2[1:positives]>0.05)
+  c(FP1,FN1,FP2,FN2,FP3,FN3)
+}) 
+
+```
+
+```{r saverds,eval=FALSE, include=FALSE}
+# saveRDS(results_global,"data/results_global.rds")
```

Comment:
You did not run this, so the rows below this fail.

Please save this file (probably as 13_results_global.rds) and make sure everything works in a new R session. Try ctrl-shift-k to render it locally in a new session. If it fails, it will cause the book to fail to render.